modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxStateImpl.java

```diff
@@ -465,4 +478,257 @@
     public String toString() {
         return S.toString(IgniteTxStateImpl.class, this, "txMap", allEntriesCopy());
     }
+
+    /**
+     * Creates savepoint.
+     *
+     * @param name Savepoint ID.
+     * @param overwrite If {@code true} - already created savepoint with the same name will be replaced.
+     */
+    public void savepoint(String name, boolean overwrite) {
+        ListIterator<TxSavepoint> spIter = findSP(name);
+
+        if (spIter != null) {
+            if (overwrite)
+                releaseSP(spIter);
+            else {
+                throw new IllegalArgumentException("Savepoint \"" + name + "\" already exists. " +
+                    "Savepoints with the same name aren't available. " +
+                    "If you want to rewrite savepoints - use savepoint(name, true) method.");
+            }
+        }
+
+        createSavepoint(name);
+    }
+
+    /**
+     * Rollback this transaction to previous state. Also deletes all afterward savepoints.
+     *
+     * @param name Savepoint ID.
+     * @throws IgniteCheckedException If failed.
+     */
+    public void rollbackToSavepoint(String name, GridNearTxLocal tx) throws IgniteCheckedException {
+        assert tx != null;
+        ListIterator<TxSavepoint> spIter = findSP(name);
+
+        if (spIter == null)
+            throw new IllegalArgumentException("No such savepoint.");
```

Comment:
Should we use `IgniteCheckedException` here?

Reply:
I think we need IAE, because savepoint's name is given by user, not something internal.